src/machines/modelingMachine.ts

```diff
@@ -519,77 +522,85 @@ export const modelingMachine = setup({
-        ast,
-        pathToNode,
-        false,
-        'variableName' in distance
-          ? distance.variableIdentifierAst
-          : distance.valueAst
-      )
-      if (trap(extrudeSketchRes)) return
-      const { modifiedAst, pathToExtrudeArg } = extrudeSketchRes
-
-      store.videoElement?.pause()
-      const updatedAst = await kclManager.updateAst(modifiedAst, true, {
-        focusPath: pathToExtrudeArg,
-        zoomToFit: true,
-        zoomOnRangeAndType: {
-          range: selection.codeBasedSelections[0].range,
-          type: 'path',
-        },
-      })
-      if (!engineCommandManager.engineConnection?.idleMode) {
-        store.videoElement?.play().catch((e) => {
-          console.warn('Video playing was prevented', e)
+      ;(async () => {
+        if (!event.data) return
+        const { selection, distance } = event.data
+        let ast = kclManager.ast
+        if (
+          'variableName' in distance &&
+          distance.variableName &&
+          distance.insertIndex !== undefined
+        ) {
+          const newBody = [...ast.body]
+          newBody.splice(
+            distance.insertIndex,
+            0,
+            distance.variableDeclarationAst
+          )
+          ast.body = newBody
+        }
+        const pathToNode = getNodePathFromSourceRange(
+          ast,
+          selection.codeBasedSelections[0].range
+        )
+        const extrudeSketchRes = extrudeSketch(
+          ast,
+          pathToNode,
+          false,
+          'variableName' in distance
+            ? distance.variableIdentifierAst
+            : distance.valueAst
+        )
+        if (trap(extrudeSketchRes)) return
+        const { modifiedAst, pathToExtrudeArg } = extrudeSketchRes
+
+        store.videoElement?.pause()
+        const updatedAst = await kclManager.updateAst(modifiedAst, true, {
+          focusPath: pathToExtrudeArg,
+          zoomToFit: true,
+          zoomOnRangeAndType: {
+            range: selection.codeBasedSelections[0].range,
+            type: 'path',
+          },
         })
-      }
-      if (updatedAst?.selections) {
-        editorManager.selectRange(updatedAst?.selections)
-      }
+        if (!engineCommandManager.engineConnection?.idleMode) {
+          store.videoElement?.play().catch((e) => {
+            console.warn('Video playing was prevented', e)
+          })
+        }
+        if (updatedAst?.selections) {
+          editorManager.selectRange(updatedAst?.selections)
+        }
+      })().catch(reportRejection)
     },
-    'AST delete selection': async ({ context: { selectionRanges } }) => {
-      let ast = kclManager.ast
-
-      const modifiedAst = await deleteFromSelection(
-        ast,
-        selectionRanges.codeBasedSelections[0],
-        kclManager.programMemory,
-        getFaceDetails
-      )
-      if (err(modifiedAst)) return
+    'AST delete selection': ({ context: { selectionRanges } }) => {
+      ;(async () => {
```

Comment:
I find this async IIFE hard to read. Are these replacements temporary, is there a better way to write these that we could move towards in the future? Fine with it for now

Reply:
Agreed, couldn't we put `async` keyword on line 578?

```suggestion
      'AST delete selection': async ({ context: { selectionRanges } }) => {
```

Reply:
But yeah I'm also fine with it for now, I see this all as transitory work.

Reply:
If we can't change the calling code to accept an async function, I don't know of another way.

You could use the `toSync()` helper that I made, but TypeScript/my editor can't seem to infer which destructured parameters are available when I do that. You'd just have to know. That's why I opted for nesting the async function inside the regular one when there were parameters.

Reply:
> couldn't we put `async` keyword on line 578?

`yarn fmt` doesn't seem to like that.

Reply:
Yeah all good just leave it for now then